### PR TITLE
GenString (List, generator and generator expression, callable support)

### PR DESCRIPTION
**Pull request by ldjebran**
Hi,
This is a draft for the issue https://github.com/omaciel/pytest-fauxfactory/issues/2 :)

### CODE REVIEW

File: pytest_fauxfactory.py [hunk 1]
```diff
@@ -29,6 +32,27 @@ def gen_string(items=None, str_type=None, *args, **kwargs):
...
+            name = fauxfactory.gen_choice(STRING_TYPES)
+        if args is None:
+            args = ()
+        if kwargs is None:
+            kwargs = {}
+
+        # todo type and value inspections on name args kwargs
+        self.name = name
+        self.args = args
+        self.kwargs = kwargs
```
Comment (mfalesni) [inline pytest_fauxfactory.py]: For this use case, `attrs` PyPI library may be useful.

Comment (ldjebran) [reply pytest_fauxfactory.py]: Hi @mfalesni , thanks for looking this is simply a draft for the moment that show the main idea, if it's OK, than we can investigate should we use third party library for this , looked at attrs nice lib to have  :).

**Comment by ldjebran** [comment]
Added callable support

**Comment by ldjebran** [comment]
@mfalesni @omaciel think has finished with the idea implementation :) what do you think?

File: pytest_fauxfactory.py [hunk 2]
```diff
@@ -15,18 +20,42 @@
...
 
-def gen_string(items=None, str_type=None, *args, **kwargs):
-    """Generate a new string type."""
-    item = 0
+class GenString(object):
+    """fauxfactory string generator helper"""
+    __slots__ = ['str_type', 'args', 'kwargs']
+
+    def __init__(self, str_type, *args, **kwargs):
+        if str_type is None:
```
Comment (omaciel) [inline pytest_fauxfactory.py]: How about a test to show that when `str_type is None` that the value associated with it is `in STRING_TYPES`?

File: pytest_fauxfactory.py [hunk 3]
```diff
@@ -15,18 +20,42 @@
...
+            str_type = fauxfactory.gen_choice(STRING_TYPES)
+
+        if (not isinstance(str_type, six.string_types)
```
Comment (omaciel) [inline pytest_fauxfactory.py]: How about a test that does pass a string but that is not callable? It should expect a `ValueError` exception.

File: pytest_fauxfactory.py [hunk 4]
```diff
@@ -15,18 +20,42 @@
...
+                and not callable(str_type)):
+            raise ValueError('str_type must be a string or callable')
 
-    if items is None:
-        items = 1
-    if str_type is None:
-        str_type = fauxfactory.gen_choice(STRING_TYPES)
+        if (isinstance(str_type, six.string_types)
```
Comment (omaciel) [inline pytest_fauxfactory.py]: Here's another test waiting to be written :)

File: pytest_fauxfactory.py [hunk 5]
```diff
@@ -48,5 +77,23 @@ def pytest_generate_tests(metafunc):
...
+        if isinstance(str_type, GenString):
+            data = str_type(items=items)
+        elif isinstance(str_type, (tuple, list)):
+            # expect that the list is a list of GenString
+            data = chain.from_iterable(
+                gn(items=items)
+                for gn in str_type
+                if isinstance(gn, GenString)
+            )
+            if not data:
```
Comment (omaciel) [inline pytest_fauxfactory.py]: Another test waiting to be written :)

File: pytest_fauxfactory.py [hunk 6]
```diff
@@ -48,5 +77,23 @@ def pytest_generate_tests(metafunc):
...
+                raise Exception('no GenString found')
+        elif isinstance(str_type, six.string_types) or callable(str_type):
```
Comment (omaciel) [inline pytest_fauxfactory.py]: Another test waiting to be written :)

File: tests/test_pytest_fauxfactory.py [hunk 7]
```diff
@@ -1,6 +1,21 @@
...
 """Tests the `gen_string` mark."""
+import contextlib
+
 import pytest
+from pytest_fauxfactory import GenString
+import fauxfactory
+import six
+
+
+@contextlib.contextmanager
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply tests/test_pytest_fauxfactory.py]: this is for negative tests that should raise exceptions

**Comment by ldjebran** [comment]
@omaciel updated, and made some changes :)

**Comment by ldjebran** [comment]
After discussion, we have decided to split the functionality as per the created issues https://github.com/omaciel/pytest-fauxfactory/issues/5 and https://github.com/omaciel/pytest-fauxfactory/issues/6

**Pull request closed without merging**